File: scripts/fetch-bank-data.js
/**
 * teraren API から全銀行・全支店を取得し、assets/bank-data.js を生成する。
 * 実行: node scripts/fetch-bank-data.js
 * 出力: assets/bank-data.js (window.BANK_DATA に全行・全支店を格納)
 */
const fs = require("fs");
const path = require("path");

const BANKS_URL = "https://bank.teraren.com/banks.json?page=1&per=2000";
const BRANCH_URL = (code) => `https://bank.teraren.com/banks/${code}/branches.json?page=1&per=2000`;
const OUT_PATH = path.resolve(__dirname, "..", "assets", "bank-data.js");

const DELAY_MS = 300;
const sleep = (ms) => new Promise((r) => setTimeout(r, ms));

async function fetchJSON(url, retries = 3) {
  for (let i = 0; i < retries; i++) {
    try {
      const res = await fetch(url);
      if (!res.ok) throw new Error(`HTTP ${res.status} for ${url}`);
      return await res.json();
    } catch (e) {
      if (i === retries - 1) throw e;
      console.warn(`  retry ${i + 1}/${retries} for ${url}: ${e.message}`);
      await sleep(1000);
    }
  }
}

(async () => {
  console.log("Fetching banks list...");
  const banksList = await fetchJSON(BANKS_URL);
  console.log(`  Found ${banksList.length} banks.`);

  const result = [];
  for (let i = 0; i < banksList.length; i++) {
    const b = banksList[i];
    const code = String(b.code || b.bank_code || "").padStart(4, "0");
    const name = b.name || b.bank_name || "";
    process.stdout.write(`\r  [${i + 1}/${banksList.length}] ${code} ${name}...`);

    let branches = [];
    try {
      const rawBranches = await fetchJSON(BRANCH_URL(code));
      branches = (rawBranches || []).map((br) => {
        const brCode = String(br.code || br.branch_code || "").padStart(3, "0");
        return { code: brCode.slice(-3), name: br.name || br.branch_name || "" };
      });
    } catch (e) {
      console.warn(`\n  WARN: failed to fetch branches for ${code} ${name}: ${e.message}`);
    }

    result.push({ code, name, branches });
    if (i < banksList.length - 1) await sleep(DELAY_MS);
  }

  console.log(`\nWriting ${result.length} banks to ${OUT_PATH}...`);

  const js =
    "/**\n" +
    " * 全銀行・全支店の静的データ (teraren API から自動生成)\n" +
    " * 生成日: " + new Date().toISOString().slice(0, 10) + "\n" +
    " * 再生成: node scripts/fetch-bank-data.js\n" +
    " */\n" +
    "(function(g){g.BANK_DATA=" +
    JSON.stringify(result) +
    ";})(typeof window!==\"undefined\"?window:this);\n";

  fs.writeFileSync(OUT_PATH, js, "utf-8");
  const sizeMB = (Buffer.byteLength(js, "utf-8") / 1024 / 1024).toFixed(2);
  console.log(`Done! ${result.length} banks, file size: ${sizeMB} MB`);
})();

File: scripts/fetch-bank-data.py
"""
zengin-code/source-data (GitHub) から全銀行・全支店データを取得し
assets/bank-data.js を生成する。

実行: python scripts/fetch-bank-data.py
出力: assets/bank-data.js (window.BANK_DATA に全行・全支店を格納)
"""
import json, os, sys, io, zipfile, urllib.request
from datetime import date

REPO_ZIP = "https://github.com/zengin-code/source-data/archive/refs/heads/master.zip"
SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
OUT_PATH = os.path.join(SCRIPT_DIR, "..", "assets", "bank-data.js")

def main():
    print("Downloading zengin-code source-data repo (zip)...")
    req = urllib.request.Request(REPO_ZIP, headers={"User-Agent": "bank-data-fetcher/1.0"})
    with urllib.request.urlopen(req, timeout=60) as resp:
        zip_bytes = resp.read()
    print(f"  Downloaded {len(zip_bytes) / 1024:.0f} KB")

    zf = zipfile.ZipFile(io.BytesIO(zip_bytes))
    prefix = "source-data-master/data/"

    banks_path = prefix + "banks.json"
    with zf.open(banks_path) as f:
        banks_dict = json.loads(f.read().decode("utf-8"))
    print(f"  Found {len(banks_dict)} banks")

    branch_files = [n for n in zf.namelist() if n.startswith(prefix + "branches/") and n.endswith(".json")]
    branch_map = {}
    for bf in branch_files:
        bank_code = os.path.splitext(os.path.basename(bf))[0]
        with zf.open(bf) as f:
            branch_map[bank_code] = json.loads(f.read().decode("utf-8"))

    result = []
    for code in sorted(banks_dict.keys()):
        b = banks_dict[code]
        name = b.get("name", "")
        branches_raw = branch_map.get(code, {})
        branches = []
        for br_code in sorted(branches_raw.keys()):
            br = branches_raw[br_code]
            c = str(br.get("code", br_code)).zfill(3)[-3:]
            branches.append({"code": c, "name": br.get("name", "")})
        result.append({"code": code, "name": name, "branches": branches})

    total_branches = sum(len(b["branches"]) for b in result)
    print(f"  Total: {len(result)} banks, {total_branches} branches")

    data_json = json.dumps(result, ensure_ascii=False, separators=(",", ":"))
    js = (
        "/**\n"
        " * 全銀行・全支店の静的データ (zengin-code/source-data から自動生成)\n"
        f" * 生成日: {date.today().isoformat()}\n"
        f" * 銀行数: {len(result)}, 支店数: {total_branches}\n"
        " * 再生成: python scripts/fetch-bank-data.py\n"
        " * ライセンス: MIT (zengin-code)\n"
        " */\n"
        '(function(g){g.BANK_DATA=' + data_json + ';})(typeof window!=="undefined"?window:this);\n'
    )

    os.makedirs(os.path.dirname(OUT_PATH), exist_ok=True)
    with open(OUT_PATH, "w", encoding="utf-8") as f:
        f.write(js)

    size_mb = len(js.encode("utf-8")) / 1024 / 1024
    print(f"Done! {len(result)} banks, {total_branches} branches, file: {size_mb:.2f} MB")
    print(f"Output: {OUT_PATH}")

if __name__ == "__main__":
    main()
